src/node/mini_miner.h

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_NODE_MINI_MINER_H
+#define BITCOIN_NODE_MINI_MINER_H
+
+#include <txmempool.h>
+
+#include <memory>
+#include <optional>
+#include <stdint.h>
+
+namespace node {
+
+// Container for tracking updates to ancestor feerate as we include ancestors in the "block"
+class MiniMinerMempoolEntry
+{
+    const CAmount fee_individual;
+    const CTransaction tx;
+    const int64_t vsize_individual;
+
+// This class must be constructed while holding mempool.cs. After construction, the object's
+// methods can be called without holding that lock.
+public:
+    CAmount fee_with_ancestors;
+    int64_t vsize_with_ancestors;
+    explicit MiniMinerMempoolEntry(CTxMemPool::txiter entry) :
+        fee_individual{entry->GetModifiedFee()},
+        tx{entry->GetTx()},
+        vsize_individual(entry->GetTxSize()),
+        fee_with_ancestors{entry->GetModFeesWithAncestors()},
+        vsize_with_ancestors(entry->GetSizeWithAncestors())
+    { }
+
+    CAmount GetModifiedFee() const { return fee_individual; }
+    CAmount GetModFeesWithAncestors() const { return fee_with_ancestors; }
+    int64_t GetTxSize() const { return vsize_individual; }
+    int64_t GetSizeWithAncestors() const { return vsize_with_ancestors; }
+    const CTransaction& GetTx() const LIFETIMEBOUND { return tx; }
+};
+
+void UpdateForMinedAncestor(const MiniMinerMempoolEntry& ancestor, const MiniMinerMempoolEntry& descendant);
+
+// Comparator needed for std::set<MockEntryMap::iterator>
+struct IteratorComparator
+{
+    template<typename I>
+    bool operator()(const I& a, const I& b) const
+    {
+        return &(*a) < &(*b);
+    }
+};
+
+/** A minimal version of BlockAssembler. Allows us to run the mining algorithm on a subset of
+ * mempool transactions, ignoring consensus rules, to calculate mining scores. */
+class MiniMiner
+{
+    // When true, a caller may use CalculateBumpFees(). Becomes false if we failed to retrieve
+    // mempool entries (i.e. cluster size too large) or bump fees have already been calculated.
+    bool m_ready_to_calculate{true};
+
+    // Original outpoints requested
+    std::vector<COutPoint> m_requested_outpoints;
+
+    // Set once per lifetime, fill in during initialization.
+    // txids of to-be-replaced transactions
+    std::set<uint256> m_to_be_replaced;
+
+    // If multiple argument outpoints correspond to the same transaction, cache them together in
+    // a single entry indexed by txid. Then we can just work with txids since all outpoints from
+    // the same tx will have the same bumpfee. Excludes non-mempool transactions.
+    std::map<uint256, std::vector<COutPoint>> m_requested_outpoints_by_txid;
```

Comment:
This isn't a very efficient data structure, because each entry's key, the txid, is repeated in all of its vector's entries (because a `COutPoint` includes the txid). All you need is the index (`COutPoint::n`) So this could be:
```
    std::map<uint256, std::vector<uint32_t>> m_requested_outpoints_by_txid;
```
although the name would be wrong, it's no longer "requested outpoints". Elsewhere in the code where you need a `COutPoint` (I think there are only two places), you can construct it from the key (txid) and this index.